### ISSUE TITLE
Unique constraint not enforced on delete

### ISSUE THREAD

**Issue by mtornwall** [Starcounter/Starcounter#3546]
See the code below for a concise description of the problem.

This problem mostly manifests itself on reload, where the constraint *is* enforced. We end up in the awkward situation of having a working database that, when unloaded, cannot be reloaded.

Frankly I'm not sure what behavior I can reasonably expect from this. Clearly the delete cannot fail, doing so could give rise to "immortal" objects under some circumstances (e.g. referential cycle combined with the right indices).

~~~
namespace IndexTest
{
    using Starcounter;

    [Database] public class Foo { public Bar Bar; }
    [Database] public class Bar { }

    class Program
    {
        static void Main()
        {
            Db.SlowSQL("CREATE UNIQUE INDEX FooIndex ON IndexTest.Foo(Bar)");

            Db.Transact(() =>
            {
                var bar = new Bar();
                var foo1 = new Foo();
                var foo2 = new Foo() { Bar = bar };

                // Uh-oh...
                bar.Delete();

                // At this point we have:
                // foo1.Bar == null
                // foo2.Bar == null
                // In violation of FooIndex.
            });
        }
    }
}
~~~

**Comment by k-rus** [comment]
Thank you for reporting.

I remember there was discussion if NULL values should be treated unique (as in SQL standard) or not. @eriohl ?

@mtornwall 
If you don't define unique index, then you should be able to reload the database, shouldn't you?

**Comment by mtornwall** [comment]
> If you don't define unique index, then you should be able to reload the database, shouldn't you?

Yes, the fix is to drop the indices, reload the database, delete the bad rows, and then recreate the indices. But I think this needs to be handled consistently. It's no good to have two conflicting behaviors.

**Comment by k-rus** [comment]
So I guess NULL values are not unique currently, so two null values violate unique constraint. However, deleting an object does not affect references to it, since it is expensive and complicated. (Objects do not know who has references to them. Thus deleting them does not update the reference value to null)

This definitely brings inconsistent behaviour described above.

cc @eriohl @dan31

**Comment by eriohl** [comment]
As I see the inconsistent behavior is that the unload changes the value stored in the column to undefined even if a value is stored (id of the target record). You are allowed to set a reference to a value that doesn't exist by design. The unload should then not change the value of the column because the target does not exist.

**Comment by mtornwall** [comment]
@eriohl So you're saying that:

~~~
INSERT INTO "Starcounter"."Raw"."IndexTest"."Foo"(__id,"Bar")VALUES(object 2,NULL),(object 3,NULL)
~~~

Should instead be:

~~~
INSERT INTO "Starcounter"."Raw"."IndexTest"."Foo"(__id,"Bar")VALUES(object 2,NULL),(object 3,object 1)
~~~

With `object 1` being a reference to the `Bar` that was deleted?

**Comment by eriohl** [comment]
I suggest that the source of the inconsistency is that it is not as in the second example yes.

**Comment by mtornwall** [comment]
Any news on this?

**Comment by k-rus** [comment]
@mtornwall 
As I see you can suggest what kind of tool you want to have to help you to deal with this data inconsistency. I suspect that application code allowed deleting data without dealing with data referencing the deleted. Can you describe how the situation happened in your case that we understand better your use case?

We cannot really know what actually to do with such problem in general. But we can see if we can provide some tools.

**Comment by mtornwall** [comment]
The root of this particular problem is indeed my application's faulty deletion logic. However, this wouldn't in itself have caused any significant issues as I could easily have cleaned up the orphaned objects later.

If the unload/reload process is to be the officially sanctioned mechanism for transferring data to a larger database, or for upgrading to a new Starcounter version, I think it's imperative that the original database and the reloaded database be for all practical purposes interchangeable.

Clearly the databases are not interchangeable if the unload process transforms dangling references to `NULL` values. So I vote for the solution indicated by @eriohl.

**Comment by k-rus** [comment]
> Clearly the databases are not interchangeable if the unload process transforms dangling references to NULL values.

This is not how it works. The unload process gets the null value and is not aware that it was not stored. I don't know if the transformation is implemented behind kernel API or in Level1 VMDBS.



> So I vote for the solution indicated by @eriohl.

You mean you want to get object reference, which cannot be resolved to actual object. In such case it can happen that when you load it will point to a different object.

**Comment by mtornwall** [comment]
> You mean you want to get object reference, which cannot be resolved to actual object.
> In such case it can happen that when you load it will point to a different object.

Yes, this would be my preferred behavior. If reference X points to object #Y in the original database, I want it to point to object #Y in the reloaded database as well -- regardless of which object that might be (modulo the need to respect object ID shift, if any).

Can the situation where object #Y means different things in the original database compared to the reloaded one occur in practice? I guess it could happen if the database into which we attempt to reload already contains an object #Y. But that sounds to me like it should be undefined behavior anyway. :-)

**Comment by k-rus** [comment]
> Can the situation where object #Y means different things in the original database compared to the reloaded one occur in practice? I guess it could happen if the database into which we attempt to reload already contains an object #Y. But that sounds to me like it should be undefined behavior anyway. :-)

Yes, this situation can happen, but only if object Y does not exist in the original database, but was or will be created in the new database. If Object Y existed and was unloaded then the reload will fail, since the object id for Y is occupied by the other object. Thus keeping reference to unexisting object does not look as good option for me.

**Comment by bigwad** [comment]
Could be fixed if reload creates and removes "faked" object with specified object_id.
May be achieved by putting appropriate insert and delete statements doing unload.

**Comment by k-rus** [comment]
> Could be fixed if reload creates and removes "faked" object with specified object_id.

I am not sure if it will help against automatically allocating object_id for an object, which was inserted with the given object id and then deleted. I don't know how this situation is actually treated in the kernel.

I don't really see which solution will be the best for UX. Basically, we need to help user to notice the errorness in data and then ability to fix.

**Comment by bigwad** [comment]
>I don't know how this situation is actually treated in the kernel.

Once you insert an object with specific object_id, that object_id will never be allocated by kernel.

>I don't really see which solution will be the best for UX. Basically, we need to help user to notice the errorness in data and then ability to fix.

I just wanted to point out that there is a mean to easily guarantee unload/reload safety. 
But if you want more generic solution - I see the problem is to distinguish legitimate null value from the artificial one, that starcounter returns when user code references non existing object.
I see several options. All of them break compatibility. And the last two still require changes is unload/reload.
- to force referential constraints: thus do not allow to delete an object if reference still exists to the object being deleted. Most restrictive and expensive both in development and runtime.
- to return special object if user code references non-existing object (not easy considering static types).
- to throw an exception if user code references non-existing object.
